File: src/Orbit.cc
// -*- LSST-C++ -*-

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
/* File: Orbit.cc
 * Author: Matthew Cleveland
 * Purpose: Implements Orbit object
 */

#include "lsst/mops/Orbit.h"


namespace lsst {
    namespace mops {


/**************************************************
 * Print Orbit data
 **************************************************/
void Orbit::print()
{
  std::cout << "perihelion: " << perihelion << "  eccentricity: "
	    << eccentricity << "  inclination: " << inclination 
	    << "  perihelion arg: " << perihelionArg << "  longitude: " 
	    << longitude << "  perihelion time: " << perihelionTime 
    	    << "  equinox: " << equinox 
	    << "  orbitID: " << orbitID << std::endl;
}




/**************************************************
 * Populate all data fields from formatted string
 **************************************************/
void Orbit::populateOrbitFromString(std::string values, int index)
{
  std::istringstream iss(values);

  //iss.exceptions(std::ifstream::failbit | std::ifstream::badbit);
  
  //try{
  iss >> perihelion;
  iss >> eccentricity;
  iss >> inclination;
  iss >> perihelionArg;
  iss >> longitude;
  iss >> perihelionTime;
  iss >> equinox;
  
  //check for optional orbitID column, if not there, use
  //line number
  if(!iss.eof()){
    iss >> orbitID;
  }
  else{
    orbitID = index;
  }
  
  /*  }
      catch (...){
      throw LSST_EXCEPT(BadParameterException, 
      "Badly-formatted Orbit string\n");
      }*/
}



/**************************************************
 * SETTERS
 **************************************************/
void Orbit::setPerihelion(double _p)
{
  perihelion = _p;
}

void Orbit::setEccentricity(double _e)
{
  eccentricity = _e;
}

void Orbit::setInclination(double _i)
{
  inclination = _i;
}

void Orbit::setPerihelionArg(double _p)
{
  perihelionArg = _p;
}

void Orbit::setLongitude(double _l)
{
  longitude = _l;
}

void Orbit::setPerihelionTime(double _p)
{
  perihelionTime = _p;
}

void Orbit::setEquinox(double _e)
{
  equinox = _e;
}

void Orbit::setOrbitID(double _o)
{
  orbitID = _o;
}



/**************************************************
 * GETTERS
 **************************************************/
double Orbit::getPerihelion() const
{
  return perihelion;
}

double Orbit::getEccentricity() const
{
  return eccentricity;

}

double Orbit::getInclination() const
{
  return inclination;
}

double Orbit::getPerihelionArg() const
{
  return perihelionArg;
}

double Orbit::getLongitude() const
{
  return longitude;
}

double Orbit::getPerihelionTime() const
{
  return perihelionTime;
}

double Orbit::getEquinox() const
{
  return equinox;
}

double Orbit::getOrbitID() const
{
  return orbitID;
}





    }} //close lsst::mops

File: python/lsst/mops/daymops/Orbit.py
# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""
Class to represent an Orbit object (part of a MovingObject instance).
"""
from DayMOPSObject import DayMOPSObject



# Constants
STABLE_STATUS = {'STABLE':      'Y',
                 'UNSTABLE':    'N'}


class Orbit(DayMOPSObject):
    """
    Representation of an Orbit.
    
    This, together with the MovingObject class, fully describe the subset of the
    MovingObject table of interset to MOPS.
    """
    def __init__(self, 
                 q=None, 
                 e=None, 
                 i=None, 
                 node=None, 
                 argPeri=None, 
                 meanAnom=None,
                 timePeri=None, 
                 epoch=None, 
                 src=[],
                 orbFitResidual=None,
                 orbFitChi2=None,
                 classification=None,
                 stablePass=STABLE_STATUS['UNSTABLE'],
                 moid1=None,
                 moid2=None,
                 moidLong1=None,
                 moidLong2=None):
        """
        a (AU)                              (only for Keplerian elements)
        q (AU)                              (only for cometary elements)
        e
        i (deg)
        node (deg)
        argPeri (deg)
        m\meanAnom (deg)                    (only for Keplerian elements)
        timePeri (TAI MJD)                  (only for cometary elements)
        epoch: orbit epoch (TAI MJD)
        src: 21 element array (covariance matrix in diagonal form).
        """
        super(Orbit, self).__init__()
        
        self._q = q
        self._e = e
        self._i = i
        self._node = node
        self._argPeri = argPeri
        self._meanAnom = meanAnom
        self._timePeri = timePeri
        self._epoch = epoch
        self._src = None
        self.setSrc(src)
        
        self._orbFitResidual = orbFitResidual
        self._orbFitChi2 = orbFitChi2
        self._classification = classification
        self._stablePass = stablePass
        self._moid1 = moid1
        self._moid2 = moid2
        self._moidLong1 = moidLong1
        self._moidLong2 = moidLong2
        return

    def __str__(self):
        return('(%s, %s,%s, %s, %s, %s, %s, %s, %s)'\
               % tuple([str(x) for x in (self._q, self._e, self._i, self._node, 
                        self._m, self._timePeri, self._argPeri, self._epoch)]))

    def setSrc(self, src):
        """
        If all elements of the covariance list are not None, then cast that
        list into a numpy.array. Return the casted array or None in case the
        covariance is invalid (i.e. has null elements).
        """
        self._src = []
        if(not None in src):
            self._src = [float(e) for e in src]
        return




